### ISSUE TITLE
[BUG] Training Plugin - Analyser and Effekt tab

### ISSUE THREAD

**Issue by Sturzbach** [akasolace/HO#539]
**Describe the bug**
There is a bug, which invalidates trainings data since Update to Version 3.0 i Think.
In a old Database everything is alright and the Data are shown with correct entries for Season and week. But then I import data now, the Acutal data are shown as of the season before.
This is the fact in Tab Training / Analyse and Effect.

The result is i think it does not caslculate the training effect correctly in this tab.
In main player attribute it seems to b correctly.

**To Reproduce**
I tokk my database from before the update and there it shows data for my players for Season 62 (Scotland  IV.31).
Then i have imported now data from today, int the traings analyser tab and effect tab i got data shown named season 61.
Also effect is empty although i had 3 Skillups on thursday and week 16 is not shown.

**Screenshots**
Good data from 21.5.2020
![grafik](https://user-images.githubusercontent.com/59139029/83326368-d8c5c580-a273-11ea-9f6d-22dfbcca0088.png)

Bad Data after import from today
![grafik](https://user-images.githubusercontent.com/59139029/83326505-e3cd2580-a274-11ea-9d4e-abf9637868bd.png)

**Platform information:**
Windows 10 Home 1909
Version 3.000

**Comment by akasolace** [comment]
@wsbrenk  I assigned it to you as you are the most experienced with that part of the code.
I let you check whether or not we have an issue here and what should be the appropriate labels and milestone...

**Comment by wsbrenk** [comment]
currently i could not reproduse this effect, but i remember that i saw it some days ago. i guess it is a week 16 effect.

**Comment by akasolace** [comment]
ok lets look again at this one on week 16 then ...

**Comment by wsbrenk** [comment]
OldTrainingManager uses trainings date to initialize PastSkillChange object. this uses HTCalendarFactory.getHTSeason(...,useTrainingCalendar=false) to set HtSeason and HTWeek. Don't know if using the economyCalendar could explain this problem.

**Comment by akasolace** [comment]
@wsbrenk sorry but I am not familiar at all with that part of the code... but I don't really understand why a function getHTSeason would return different values for HtSeason and HTWeek whether or not useTrainingCalendar is set to True

**Comment by akasolace** [comment]
@wsbrenk for a different reason I was considering a refactoring of HTCalendar and HTCalendarFactory  ( #678).  I am planning to do that for v5.0.  It would be nice to have your inputs for what you would need for the training manager. 
Also:
   - is OldTrainingManager still used somehere ?  
   - Are you still planing to have a look at this issue during week 16?  
   - if issue is confirmed during week 16, you will let me know if we need to push this one to v5.0

**Comment by wsbrenk** [comment]
> * is OldTrainingManager still used somehere ?

yes, it is used in the training effect panel and in the player's training history view

> * Are you still planing to have a look at this issue during week 16?

yes, that is the plan. tomorrow i have to start a holiday trip to the north of germany, but i will smuggle the notebook into my luggage...

> * if issue is confirmed during week 16, you will let me know if we need to push this one to v5.0

i hope i will met the interval when the error is reproducable and if so, he won't have a chance to survive

**Comment by akasolace** [comment]
>>is OldTrainingManager still used somehere ?

>yes, it is used in the training effect panel and in the player's training history view

is the plan to merge them at some point? just that the name OldTrainingManager make it sounds marked as deprecated ...

>> Are you still planing to have a look at this issue during week 16?

> yes, that is the plan. tomorrow i have to start a holiday trip to the north of germany, but i will smuggle the notebook into my luggage...

>> if issue is confirmed during week 16, you will let me know if we need to push this one to v5.0

> i hope i will met the interval when the error is reproducable and if so, he won't have a chance to survive

ok great ... also if you have requirements or idea of improvements regarding HTCalendar don't hesitate to list them in #678, I will consider them when I reimplement the class (or maybe I will have a smooth approach and create first a NewHTCalendar and mark the other one as deprecated) 
In the meantime, enjoy your holiday!

**Comment by wsbrenk** [comment]
the error was observed on may 30th. i think that was the date, when season flipped from 74 to 75.

EffectDAO uses economy calendar to calculate the HTSeason ->

   public static HTCalendar createEconomyCalendar() {
    	HOModel model = HOVerwaltung.instance().getModel();
    	final Calendar calMark = Calendar.getInstance();
    	if (model == null || model.getXtraDaten() == null || model.getXtraDaten().getEconomyDate() == null) {
    		return new HTCalendar();
    	}
    	calMark.setTimeInMillis(model.getXtraDaten().getEconomyDate().getTime());
    	final HTCalendar calendar = new HTCalendar();
    	calendar.initialize(calMark);
    	calendar.setTime(model.getBasics().getDatum());

    	final int correction = calendar.getHTSeason() - model.getBasics().getSeason();

    	calendar.setSeasonCorrection(correction);
    	return calendar;
    }

i think the correction could be the bug, if the economy season does not fit to the season stored in the basics table, which is from worldetails.xml file, i think. difference of economy and training date in germany is either nearly one day or six days depending on the download date.

if that is true, then i see 2 possibilities to repair that.
- use the correct calendar to calculate the ht season (same as in world details)
- get the htseason from world details (basics table) instead of calculating it on our own

**Comment by wsbrenk** [comment]
createTrainingCalendar calculates the correction exactly the same way as the economy calendar. One of them must be wrong.

**Comment by akasolace** [comment]
> get the htseason from world details (basics table) instead of calculating it on our own

It might be best to avoid parsing this file everytime, also it is best to have correct week/season in HO even if user does not download new data....

> use the correct calendar to calculate the ht season (same as in world details)

this is my idea to fix that HTCalendar class, I now how to make reliable function to calculate HTSeason and HTWeek whic start on Monday. Now I need to check where to get the time for economy and traininng update per league (I hope it is listeed in one xml file). Then I will additional function that be useful for the training manager or other parts of HO

**Comment by wsbrenk** [comment]
economy date is NOT the date worldetails flips the season:

SELECT  datum, saison, seriesmatchdate, trainingdate, economydate FROM basics, XTRADATA WHERE basics.hrf_id=XTRADATA.hrf_id order by datum gives:


DATUM | SAISON | SERIESMATCHDATE | TRAININGDATE | ECONOMYDATE
-- | -- | -- | -- | --
2020-05-30 16:01:48.000000 | 74 | 2020-05-30 18:00:00.000000 | 2020-06-05 00:30:00.000000 | 2020-06-06 00:10:00.000000
2020-06-01 11:41:31.000000 | 75 | 2020-06-06 18:00:00.000000 | 2020-06-05 00:30:00.000000 | 2020-06-06 00:10:00.000000

**Comment by akasolace** [comment]
> economy date is NOT the date worldetails flips the season it is the match date!!!

I would expect worldetails to switch season according to official calendar it means at the end of week 16, on Monday

**Comment by wsbrenk** [comment]
> this is coming from an xml file I guess ...

if i investigate correctly, it's from worlddetails

**Comment by wsbrenk** [comment]
for the season correction calculation we need the exact date time info, when worlddetails switches the season number. 

the correction calculation could be done static, with a calendar matching this exact timestamp. not in each calendar with another fix point.

**Comment by akasolace** [comment]
> for the season correction calculation we need the exact date time info, when worlddetails switches the season number.

In all logic it should happen Monday midnight

**Comment by akasolace** [comment]
> > this is coming from an xml file I guess ...
> 
> if i investigate correctly, it's from worlddetails

yes you are right, it is from this table, I had quick look and I don't see when this is udpated. Do you know if this table is updated every time user downloan new data (i.e. world details downloded every time) ?

**Comment by akasolace** [comment]
I updated #679 with new info

**Comment by wsbrenk** [comment]
> In all logic it should happen Monday midnight

i don't think so. i added my downloads around the flip from 73 to 74. there you can see, that at 22 o'clock the season is already 74. i don't think that sweden has a time zone 2 hours in advance of MET

**Comment by akasolace** [comment]
> season 73 i see at 19:52 after last match date (it is the match date!)

@wsbrenk  you are right, I asked to HT devs cause this might be a bug, the documentation said that HTweek flips Sunday midnight, so I would have expected the same for Season. For my info, you need the HTSeason flip for which part of the code? I guess it has not to do with the training ......

**Comment by wsbrenk** [comment]
we need it to solve this issue. i think this is the reason, why in the time between the seasons the training effect panel will show wrong season number for at least one day.

**Comment by akasolace** [comment]
@wsbrenk I am not sure of what are the need on your side, but I would suggest the following:

XTRATABLE:  store seasonoffset, training/ecoonomy/match League day and hour

a new class HTCalendarNew that will have methods

 - fromGregorianToHTCalendar(date) -> (Season, Week, Next Training, Next Economy) 

Week: official rule start Monday 0 am 
Season: waiting for confirmation of HTDevs

In that case, I don't think we would need anymore to store the Season in XTRAtable we will only rely on call to fromGregorianToHTCalendar.  

@wsbrenk  how does that sounds ?

**Comment by wsbrenk** [comment]
> @wsbrenk how does that sounds ?

your plans concerning refactoring HTCalendar really sounds good. (But i would not keep the old version. I fear that it will continue to do misfortune for a long time to come.)

> In that case, I don't think we would need anymore to store the Season in XTRAtable we will only rely on call to fromGregorianToHTCalendar.

in my opinion we should store the whole stuff, we could get from the xml files. you never know if you could use it for anything in the future. it is much more harder to reload things you missed than storing some things you don't really need. (one example i examined some days ago when i needed the match part of match events, which was missing in the database.)

**Comment by akasolace** [comment]
@wsbrenk 
I did some analysis and I think now it is finally clear to me:
   - HT week start on Monday
   - HT Season switch at the end of week 16
   - worlddetails HTSeason is misleading (might indicate which season is the coming game but I think we can discard it all together).
   - as a result most country have league game from week 1 to 14 but some county (e.g. Brazil) have league game from week 2 to 15.

Now we need to clarify what we want to see in Analyser and effect tab: do we want global week in HT time, do we want to support local time accros HO?, Do we want it to be simply the corresponding league round?

in case of (1), we just need to look at the game played during global week 1 (easy)  , in case of (2) we need for let say week 3, the third league game and a friendly/cup game between the 3rd and the 4th game.  In that case we could imagine a function in HTCalendar that returns a min/max dates something like getStartEndTrainingWeek(Season S, Week W, League L) ???

Hence I think there are still some work before we can completly close this issue, don't you think?

**Comment by wsbrenk** [comment]
> Hence I think there are still some work before we can completly close this issue, don't you think?

it depends - this issue was that for some days at the end of the season the user will see a complete wrong ht season information in training effect table. the changes of this pr will hopefully repair this.

Another thing is, that HTCalendarFactory could need a refactoring - i don't see any reason, why there should be different calendars for economy or training week. As you mention in the other issue, that would be a greater thing and needs a lot of testing of all the parts where HTCalendar is used. if you will get ready with that for V4, this pr will be unnecessary.

**Comment by akasolace** [comment]
ok let's merge this one hoping it fixes the issue and have a deeper analysy on HTCalendar refactoring

**Comment by akasolace** [comment]
should be fixed by #683 => to be closed after testing

**Comment by wsbrenk** [comment]
![image](https://user-images.githubusercontent.com/11821502/93666445-a4725100-fa7e-11ea-846f-e92d69e164fa.png)

**Comment by wsbrenk** [comment]
HO4 before download;-(

![image](https://user-images.githubusercontent.com/11821502/93666524-1e0a3f00-fa7f-11ea-8ba4-b5523ecebc26.png)

**Comment by wsbrenk** [comment]
now HO4 display is fine!
